readme.md
<p align="center"><img width="50%" src="posts/img/title.png" /></p>

Here are my personal deep learning notes. I've written this cheatsheet for keep track my knowledge
but you can use it as a guide for learning deep learning aswell.


<!------------------------ Dataset ------------------------>
<table>
  <tr>
    <th rowspan="4" width="150"><h3>🗂</br>Dataset</h3></th>
        <th align="left"><a href="#balance-the-data">             Balance the data </b>         </a></th> </tr>
  <tr>  <th align="left"><a href="#split-in-train-and-validation">Split in train and validation </a></th> </tr>
  <tr>  <th align="left"><a href="#normalization">                Normalization                 </a></th> </tr>
  <tr>  <th align="left"><a href="#data-augmentation">            Data augmentation             </a></th> </tr>
</table>
  
<!------------------------ Layers ------------------------>
<table>
  <tr>   <th rowspan="5" width="150"><h3>🧠</br>Model</h3></th>
         <th align="left"><a href="#activation-function">   Activation function  </a></th></tr>
  <tr>   <th align="left"><a href="#weight-initialization"> Weight initialization </a></th></tr>
  <tr>   <th align="left"><a href="#batch-normalization">   Batch normalization  </a></th></tr>
  <tr>   <th align="left"><a href="#self-attention">        Self-attention       </a></th></tr>
</table>
  
<!------------------------ Loss ------------------------>
<table>
  <tr>  <th rowspan="3" width="150"><h3>📉</br>Loss</h3></th>
        <th align="left"><a href="#loss-function">  Loss Function  </a></th></tr>
  <tr>  <th align="left"><a href="#weight-penalty"> Weight Penalty </a></th></tr>
  <tr>  <th align="left"><a href="#label-tricks"> Label Tricks </a></th></tr>
</table>
  
<!------------------------ Train ------------------------>
<table>
  <tr> <th rowspan="4" width="150"><h3>🔥</br>Train</h3></th>
       <th align="left"><a href="#optimizer">        Optimizer     </a></th></tr>
  <tr> <th align="left"><a href="#learning-rate">    Learning Rate </a></th></tr>
  <tr> <th align="left"><a href="#batch-size">       Batch size    </a></th></tr>
  <tr> <th align="left"><a href="#number-of-epochs"> Num epochs    </a></th></tr>
</table>

<!------------------------ Avoid overfitting ------------------------>
<table>
  <tr>
    <th rowspan="5" width="150"><h3>🧐</br>Avoid</br>overfitting</h3><h5>(Try in that order)</h5></th>
         <td><a href="#get-more-data">          1. Get more data            </a></td></tr>
  <tr>   <td><a href="#data-augmentation">      2. Data augmentation        </a></td></tr>
  <tr>   <td><a href="#regularization">         3. Regularization           </a></td></tr>
  <tr>   <td><a href="#reduce-model-complexity">4. Reduce Reduce complexity </a></td></tr>
  <tr>   <td><a href="#ensemble">               5. Ensemble                 </a></td></tr>
</table>

<!------------------------ Train faster ------------------------>
<table>
  <tr>
    <th rowspan="5" width="150"><h3>🕓</br>Train</br>faster</h3></th>
         <td><a href="#transfer-learning">  Transfer learning   </a></td></tr>
  <tr>   <td><a href="#batch-normalization">Batch Normalization </a></td></tr>
  <tr>   <td><a href="#precomputation">     Precomputation      </a></td></tr>
  <tr>   <td><a href="#half-precision">     Half precision      </a></td></tr>
  <tr>   <td><a href="#multiple-gpus">      Multiple GPUs       </a></td></tr>
</table>


<!------------------------ Applications ------------------------>
<table>
  <tr> <th rowspan="5" width="150"><h3>🤖</br>Applications</h3><h5>(External repos)</h5></th>
       <th><a href="https://github.com/javiabellan/vision"> Vision                 </a></th> </tr>
  <tr> <th><a href="https://github.com/javiabellan/nlp">    NLP                    </a></th> </tr>
  <tr> <th><a href="https://github.com/javiabellan/audio">  Audio                  </a></th> </tr>
  <tr> <th><a href="https://github.com/javiabellan/tabular">Tabular                </a></th> </tr>
  <tr> <th><a href="https://github.com/javiabellan/rl">     Reinforcement Learning </a></th> </tr>
</table>


<!------------------------ Computer ------------------------>
<table>
  <tr> <th rowspan="4" width="150"><h3>🖥️</br>Computer</h3></th>
       <th align="left"><a href="/posts/0-setup/hardware.md"> Hardware </a></th></tr>
  <tr> <th align="left"><a href="/posts/0-setup/software.md"> Software </a></th></tr>
  <tr> <th align="left"><a href="/posts/0-setup/jupyter.md">  Jupyter  </a></th></tr>
  <tr> <th align="left"><a href="/posts/0-setup/kaggle.md">   Kaggle   </a></th></tr>
</table>
  

<!------------------------ Resources ------------------------>
<table>
  <tr>
    <th rowspan="5" width="150">
      <h3>
        <a href="#resources">Resources</a>
      </h3>
    </th>
  </tr>
</table>

---

<h1 align="center">🗂 Dataset</h1>

## Balance the data
- **Fix it in the dataloader** [`WeightedRandomSampler`](https://pytorch.org/docs/stable/data.html#torch.utils.data.WeightedRandomSampler)
- **Subsample majority class**. But you can lose important data.
- **Oversample minority class**. But you can overfit.
- **Weighted loss function** `CrossEntropyLoss(weight=[…])`
 
## Split in train and validation
- **Training set**: used for learning the parameters of the model. 
- **Validation set**: used for evaluating model while training. Don’t create a random validation set! Manually create one so that it matches the distribution of your data. Usaully a `10%` or `20%` of your train set.
  - N-fold cross-validation. Usually `10`
- **Test set**: used to get a final estimate of how well the network works.

## Normalization
Scale the inputs to have mean 0 and a variance of 1. Also linear decorrelation/whitening/pca helps a lot. Normalization parameters are obtained only **from train set**, and then applied to both train and valid sets.
- Option 1: **Standarization** `x = x-x.mean() / x.std()` *Most used*
   1. **Mean subtraction**: Center the data to zero. `x = x - x.mean()` fights vanishing and exploding gradients
   2. **Standardize**: Put the data on the same scale. `x = x / x.std()` improves convergence speed and accuracy
- Option 2: **PCA Whitening**
  1. **Mean subtraction**: Center the data in zero. `x = x - x.mean()`
  2. **Decorrelation** or **PCA**: Rotate the data until there is no correlation anymore.
  3. **Whitening**: Put the data on the same scale. `whitened = decorrelated / np.sqrt(eigVals + 1e-5)`
- Option 3: **ZCA whitening** Zero component analysis (ZCA).
- Other options not used:
  - `(x-x.min()) / (x.max()-x.min())`: Values from 0 to 1
  - `2*(x-x.min()) / (x.max()-x.min()) - 1`: Values from -1 to 1
  
> - In case of images, the scale is from 0 to 255, so it is not strictly necessary normalize.
> - [**neural networks data preparation**](http://cs231n.github.io/neural-networks-2/#datapre)


## Data augmentation
Todo




<h1 align="center">🧠 Model</h1>

## Activation function
> [reference](https://mlfromscratch.com/activation-functions-explained)
- **Softmax**: Sigle-label classification (last layer)
- **Sigmoid**: Multi-label classification (last layer)
- **Hyperbolic tangent**:
- **ReLU**: Non-linearity compontent of the net (hidden layers) check [this paper](https://arxiv.org/pdf/1710.05941.pdf)
- **ELU**: Exponential Linear Unit. [paper](https://arxiv.org/abs/1511.07289)
- **SELU**: Scaled Exponential Linear Unit. [paper](https://arxiv.org/abs/1706.02515)
- **PReLU** or **Leaky ReLU**:
- **SERLU**:
- Smoother ReLU. Differienzable. **BEST**
  - **GeLU**: Gaussian Error Linear Units. Used in transformers. [paper](https://arxiv.org/abs/1606.08415). (2016)
  - **Swish**: `x * sigmoid(x)` [paper](https://arxiv.org/abs/1710.05941) (2017)
  - **Elish**: `xxxx` [paper](https://arxiv.org/abs/1808.00783) (2018)
  - **Mish**: `x * tanh( ln(1 + e^x) )` [paper](https://arxiv.org/abs/1908.08681) (2019)
  - **myActFunc 1** = `0.5 * x * ( tanh(x) + 1 )`
  - **myActFunc 2** = `0.5 * x * ( tanh (x+1) + 1)`
  - **myActFunc 3** = `x * ((x+x+1)/(abs(x+1) + abs(x)) * 0.5 + 0.5)`

## Weight initialization
Depends on the models architecture. Try to avoid vanishing or exploding outputs. [blog1](https://towardsdatascience.com/weight-initialization-in-neural-networks-a-journey-from-the-basics-to-kaiming-954fb9b47c79), [blog2](https://madaan.github.io/init/).
- **Constant value**: Very bad
- **Random**:
  - Uniform: From 0 to 1. Or from -1 to 1. Bad
  - Normal: Mean 0, std=1. Better
- **Xavier initialization**:  Good for MLPs with tanh activation func. [paper](http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf)
  - Uniform: 
  - Normal: 
- **Kaiming initialization**: Good for MLPs with ReLU activation func. (a.k.a. He initialization) [paper](https://arxiv.org/abs/1502.01852)
  - Uniform
  - Normal
  - When you use Kaiming, you ha to fix `ReLU(x)` equals to **`min(x,0) - 0.5`** for a correct mean (0)
- **Delta-Orthogonal initialization**: Good for vanilla CNNs (10000 layers). Read this [paper](https://arxiv.org/abs/1806.05393)


<h1 align="center">🧐 Regularization</h1>

### Dropout
During training, some **neurons** will be deactivated **randomly**. [Hinton, 2012](http://www.cs.toronto.edu/~hinton/absps/dropout.pdf), [Srivasta, 2014](https://www.cs.toronto.edu/~hinton/absps/JMLRdropout.pdf)

<p align="center"><img width="50%" src="/posts/img/dropout.png" /></p>

### Weight regularization
Weight penalty: Regularization in loss function (penalice high weights). `Weight decay` hyper-parameter usually `0.0005`.

Visually, the weights only can take a value inside the blue region, and the red circles represent the minimum. Here, there are 2 weight variables.

L1 (LASSO) |	L2 (Ridge)	| Elastic Net
-----------|--------------|------------
![wr1](/posts/img/weight-reg-lasso.png) | ![wr2](/posts/img/weight-reg-ridge.png) | ![wr3](/posts/img/weight-reg-elastic-net.png)
Shrinks coefficients to 0. Good for variable selection | **Most used**. Makes coefficients smaller | Tradeoff between variable selection and small coefficients
Penalizes the sum of absolute weights | Penalizes the sum of squared weights | Combination of 2 before
`loss + wd * weights.abs().sum()` | `loss + wd * weights.pow(2).sum()` |

### DropConnect
At training and inference, some **connections** (weights) will be deactivated **permanently**. [LeCun, 2013](http://yann.lecun.com/exdb/publis/pdf/wan-icml-13.pdf). This is very useful at the firsts layers.

<p align="center"><img width="30%" src="/img/dropconnect.jpg" /></p>

### Weight Pruning
An automatic way of DropConnect. **Iterative magnitude pruning** is iterative process of removing connections (Prune/Train/Repeat):

<p align="center"><img width="60%" src="/img/TrainLargeThenCompress.jpg" /></p>

1. Train a big model
2. Do early stopping
3. Compress model
   - **Prune**: Find the 15% of weights with the smallest magnitude and set them to zero.
   - **Train**: Then finetune the model until it reaches within 99.5% of its original validation accuracy.
   - **Repeat**: Then prune another 15% of the smallest magnitude weights and finetune.
   
At the end you can have pruned the 15%, 30%, 45%, 60%, 75%, and 90% of your original model.

Papers:
- [Deep Compression](https://arxiv.org/abs/1510.00149) (2015)
- [Train Large, Then Compress](https://arxiv.org/abs/2002.11794) (2020)
- [Neural Networks are Surprisingly Modular](https://arxiv.org/abs/2003.04881) (2020)

### Quantization and Half prcision
Usually weaight are stored in float32. Decrase the resolution is a way a regualrization and not overfit.
- **Half prcision**: Use float16.
- **Quantization**: Less bits per weight.

### Knowledge Distillation (teacher-student)

A **teacher** model teach a **student** model.

- **Smaller** student model → **faster** model.
  - Model compresion: Less memory and computation
  - To generalize and avoid outliers.
  - Used in NLP transformers.
  - [paper](https://arxiv.org/abs/1909.11723)
- **Bigger** student model is → **more accurate** model.
  - Useful when you have extra unlabeled data (kaggle competitions)
  - **1.** Train the teacher model with labeled dataset.
  - **2.** With the extra on unlabeled dataset, generate pseudo labels (soft or hard labels)
  - **3.** Train a student model on both labeled and pseudo-labeled datasets.
  - **4.** Student becomes teacher and repeat -> **2.**
  - [Paper: When Does Label Smoothing Help?](https://arxiv.org/abs/1906.02629)
  - [Paper: Noisy Student](https://arxiv.org/abs/1911.04252)
  - [Video: Noisy Student](https://youtu.be/Y8YaU9mv_us)



<h1 align="center">📉 Loss</h1>

## Loss function
- **Regression**
  - **MBE: Mean Bias Error**: `mean(GT - pred)` It could determine if the model has positive bias or negative bias.
  - **MAE: Mean Absolute Error (L1 loss)**: `mean(|GT - pred|)` The most simple.
  - **MSE: Mean Squared Error (L2 loss)**: `mean((GT-pred)²)` Penalice large errors more than MAE. **Most used**
  - **RMSE: Root Mean Squared Error**: `sqrt(MSE)` Proportional to MSE. Value closer to MAE.
  - Percentage errors:
    - **MAPE: Mean Absolute Percentage Error**
    - **MSPE: Mean Squared Percentage Error**
    - **RMSPE: Root Mean Squared Percentage Error**
- **Classification**
  - **Cross Entropy**: Sigle-label classification. Usually with **softmax**. `nn.CrossEntropyLoss`.
    - **NLL: Negative Log Likelihood** is the one-hot encoded target simplified version, see [this](https://jamesmccaffrey.wordpress.com/2016/09/25/log-loss-and-cross-entropy-are-almost-the-same/) `nn.NLLLoss()`
  - **Binary Cross Entropy**:  Multi-label classification. Usually with **sigmoid**. `nn.BCELoss`
  - **Hinge**: Multi class SVM Loss `nn.HingeEmbeddingLoss()`
  - **Focal loss**: Similar to BCE but scaled down, so the network focuses more on incorrect and low confidence labels than on increasing its confidence in the already correct labels. `-(1-p)^gamma * log(p)` [paper](https://arxiv.org/abs/1708.02002)
- **Segmentation**
  - **[Pixel-wise cross entropy](posts/img/Pixel-wise-CE.png)**
  - **IoU** (F0): `(Pred ∩ GT)/(Pred ∪ GT)` = `TP / TP + FP * FN`
  - **[Dice](posts/img/Dice.png)** (F1): `2 * (Pred ∩ GT)/(Pred + GT)` = `2·TP / 2·TP + FP * FN`
    - Range from `0` (worst) to `1` (best)
    - In order to formulate a loss function which can be minimized, we'll simply use `1 − Dice`

## Classification Metrics
Dataset with 5 disease images and 20 normal images. If the model predicts all images to be normal, its accuracy is 80%, and F1-score of such a model is 0.88
  - **Accuracy**: `TP + TN / TP + TN + FP + FN`
  - **F1 Score**: `2 * (Prec*Rec)/(Prec+Rec)`
    - **Precision**: `TP / TP + FP` = `TP / predicted possitives`
    - **Recall**: `TP / TP + FN` = `TP / actual possitives`
  - **Dice Score**: `2 * (Pred ∩ GT)/(Pred + GT)`
  - **ROC, AUC**:
  - **Log loss**:


## Label Tricks
- **Label Smoothing**: Smooth the one-hot target label
- **Mixup**: Combines pairs of examples and their labels.
  - Merge 2 samples in 1: `x_mixed = λxᵢ + (1−λ)xⱼ`
  - [Fast.ai doc](https://docs.fast.ai/callbacks.mixup.html)


<h1 align="center">🔥 Train</h1>

## Learning Rate
> How big the steps are during training.
- **Max LR**: Compute it with LR Finder (`lr_find()`)
- **LR schedule**:
  - Constant: Never use.
  - Reduce it gradually: By steps, by a decay factor, with LR annealing, etc.
    - Flat + Cosine annealing: Flat start, and then at 50%-75%, start dropping the lr based on a cosine anneal.
  - Warm restarts (SGDWR, AdamWR):
  - OneCycle: Use LRFinder to know your maximum lr. Good for Adam.

## Batch size
> Number of samples to learn simultaneously.
- **`Batch size = 1`**: Train each sample individually. (Online gradient descent) ❌
- **`Batch size = length(dataset)`**: Train the whole dataset at once, as a batch. (Batch gradient descent) ❌
- **`Batch size = number`**: Train disjoint groups of samples (Mini-batch gradient descent). ✅
  - Usually a power of 2. **`32`** or **`64`** are good values.
  - Too low: like `4`: Lot of updates. Very noisy random updates in the net (bad).
  - Too high: like `512` Few updates. Very general common updates (bad).
    - Faster computation. Takes advantage of GPU mem. But sometimes it can no be fitted (CUDA Out Of Memory)

Some people are tring to make a [batch size finder](https://forums.fast.ai/t/batch-size-finder-from-openai-implemented-using-fastai/57620) according to this [paper](https://arxiv.org/abs/1812.06162).

## Number of epochs
> Times to learn the whole dataset.
- Train until start overffiting (validation loss becomes to increase) (early stopping)

## Optimizer
> Gradient Descent methods. [reference](https://mlfromscratch.com/optimizers-explained):

|                        | Description                                        | Paper | Score |
|:-----------------------|:---------------------------------------------------|-------|-------|
| **SGD**                | Basic method. A bit slowly to get to the optimum.  |       |       |
| **SGD with Momentum**  | Speed it up with momentum, usually `mom=0.9`       |       |       |
| **AdaGrad**            | Adaptative lr                                      | 2011  |       |
| **RMSProp**            | Similar to momentum but with the gradient squared. | 2012  |       |
| **Adam**               | Combination of Momentum with RMSProp.              | 2014  | ⭐     |
| **LARS**               | Layer-wise Adaptive Rate Scaling.                  | [2017](https://arxiv.org/abs/1708.03888) ||
| **AMSGrad**            | Worse than Adam in practice. (AdamX: new verion)   | 2018  |       |
| **AdamW**              |  .                                                 | 2018  |       |
| **LAMB**               | LARS improvement.                                  | [2019](https://arxiv.org/abs/1904.00962) ||
| **NovoGrad**           |  .                                                 | [2019](https://arxiv.org/abs/1905.11286) ||
| **Lookahead**          | Is like having a buddy system to explore the loss. | [2019](https://arxiv.org/abs/1907.08610) ||
| **RAdam**              | Rectified Adam. Stabilizes training at the start.  | [2019](https://arxiv.org/abs/1908.03265) ||
| **Ranger**             | RAdam + Lookahead.                                 | 2019  | ⭐⭐⭐  |
| **RangerLars**         | RAdam + Lookahead + LARS.                          | 2019  |       |
| **Ralamb**             | RAdam + LARS.                                      | 2019  |       |
| **Selective-Backprop** | Faster training by focusing on the biggest losers. | [2019](https://arxiv.org/abs/1910.00762) ||
| **DiffGrad**           | [Solves Adam’s "overshoot" issue](https://medium.com/@lessw/meet-diffgrad-new-deep-learning-optimizer-that-solves-adams-overshoot-issue-ec63e28e01b2)                     | [2019](https://arxiv.org/abs/1909.11015) ||
| **AdaMod**             | [A new deep learning optimizer with memory](https://medium.com/@lessw/meet-adamod-a-new-deep-learning-optimizer-with-memory-f01e831b80bd)                                  | [2019](https://arxiv.org/abs/1910.12249) ||

- **SGD**: `new_w = w - lr[gradient_w]`
- **SGD with Momentum**: Usually `mom=0.9`.

history.md
<h1 align="center">History of Deep Learning</h1>

## `1943` First neuron
<img width="250" align="right" src="https://machinelearningknowledge.ai/wp-content/uploads/2019/11/McCulloch-Pitts-1.jpg">

[Walter Pitts](https://en.wikipedia.org/wiki/Walter_Pitts) and [Warren McCulloch](https://en.wikipedia.org/wiki/Warren_Sturgis_McCulloch) in their paper, *“A Logical Calculus of the Ideas Immanent in Nervous Activity”* shows the mathematical model of biological neuron. This McCulloch Pitts Neuron has very limited capability and has no learning mechanism. Yet it will lay the foundation for artificial neural network & deep learning.

## `1957` Perceptron
<img width="250" align="right" src="https://machinelearningknowledge.ai/wp-content/uploads/2019/11/Frank-Rosenblatt-Perceptron.jpg">
In his paper “The Perceptron: A Perceiving and Recognizing Automaton”, Rosenblatt shows the new avatar of McCulloch-Pitts neuron – ‘Perceptron’ that had true learning capabilities to do binary classification on it’s own. This inspires the revolution in research of shallow neural network for years to come, till first AI winter.

## `1960` First Backpropagation Model
Henry J. Kelley in his paper, “Gradient Theory of Optimal Flight Paths” shows the first ever version of continuous backpropagation model. His model is in context to Control Theory, yet it lays the foundation for further refinement in the model and would be used in ANN in future years.

## `1962` Backpropagation With Chain Rule
Backpropagation with Chain Rule
Stuart Dreyfus in his paper, “The numerical solution of variational problems” shows a backpropagation model that uses simple derivative chain rule, instead of dynamic programming which earlier backpropagation models were using. This is yet another small step that strengthens the future of deep learning.

## `1965` Birth of Multilayer Neural Network
Alexey Grigoryevich Ivakhnenko along with Valentin Grigorʹevich Lapa, creates hierarchical representation of neural network that uses polynomial activation function and are trained using Group Method of Data Handling (GMDH). It is now considered as the first ever multi-layer perceptron and Ivakhnenko is often considered as father of deep learning.

## `1969` The Fall of Perceptron
Marvin Minsky and Seymour Papert publishes the book “Perceptrons” in which they show that Rosenblatt’s perceptron cannot solve complicated functions like XOR. For such function perceptrons should be placed in multiple hidden layers which compromises perceptron learning algorithm. This setback triggers the winter of neural network research.


### References
- [Brief History of Deep Learning from 1943-2019 (Timeline)](https://machinelearningknowledge.ai/brief-history-of-deep-learning/?fbclid=IwAR09wRExjQla-ZKutxbVIDptD6vfQbaAclXIWfh8X5SjyWW61Zu76eJB2A0)
- Jürgen Schmidhuber
  - http://people.idsia.ch/~juergen/deep-learning-miraculous-year-1990-1991.html
  - http://people.idsia.ch/~juergen/2010s-our-decade-of-deep-learning.html
  - http://people.idsia.ch/~juergen/critique-honda-prize-hinton.html


javIA/train.py
###################################### IMPORTS

# Basic
import argparse
import os
import shutil
import time
import sys
import math
import numpy as np
import pathlib.Path
import cv2

# Pytorch
import torch
import torch.nn as nn
import torch.nn.parallel
import torch.backends.cudnn as cudnn
import torch.distributed as dist
import torch.optim
import torch.utils.data
import torch.utils.data.distributed
#import torch.nn.functional as F
#from torch.utils.data import DataLoader

# Torchvision
import torchvision.transforms as transforms
import torchvision.datasets as datasets
import torchvision.models as models


###################################### ARGUMENTS

parser = argparse.ArgumentParser(description='Specify some hyper parametres:')

# Training
parser.add_argument("-b",   help="Batch size",             default=64,     type=int)
parser.add_argument("-lr",  help="Learning rate",          default=0.01,   type=float)
parser.add_argument("-mo",  help="Momentum",               default=0.9,    type=float)
parser.add_argument("-wd",  help="Weight decay",           default=0.0005, type=float)
parser.add_argument("-e",   help="Number of total epochs", default=5,      type=int)
parser.add_argument("-dlb", help="Drop last batch",        action="store_true")
parser.add_argument('-r',   help='path to latest checkpoint' default='',   type=str,   metavar='PATH')

# Hardware
parser.add_argument("-cpu", help="Do not use cuda",        action="store_true")
parser.add_argument('-gpus',help='Use multiple GPUs',      action='store_true')

# Data
parser.add_argument('data', metavar='DIR', help='path to dataset') # [imagenet-folder with train and val folders]

parser.add_argument("-vp",  help="Validation percentage",  default=0.05,  type=float)
parser.add_argument('-nw',  help="Number of workers",      default=4,     type=int)

# Model
parser.add_argument('-a',   help='Model architecture: ',   default='resnet18', type=str)
parser.add_argument("-pre", help="Load pre-trained model", action="store_true")
parser.add_argument("-cp",  help="Use check_point",        action="store_true")


# Debug
parser.add_argument('-e', '--evaluate',        dest='evaluate',   action='store_true',  help='evaluate model on validation set')
parser.add_argument('--print-freq', '-p',      default=10,   type=int,   metavar='N',   help='print frequency (default: 10)')
parser.add_argument('-v',   help="Pring debug info",       action="store_true")
parser.add_argument('-d',   help="Remote debug",           action="store_true")

args = parser.parse_args()


epochs          = args.e
batch_size      = args.b
learning_rate   = args.lr
momentum        = args.mo
weight_decay    = args.wd
resume          = args.r

val_percent     = args.vp
gpu             = not args.cpu
multiple_gpus   = args.gpus
num_workers     = args.nw
drop_last_batch = args.dlb

architecture    = args.a
pretrained      = args.pre
check_point     = args.cp
verbose         = args.v

# TODO: Poner como argumentos
gpu_count       = 1
images_per_gpu  = 1

#shuffle_data    = True
CROP_SIZE       = 256


###################################### DATA

'''
In the "data" directoy:
 - The "kaggle" dir has the original data from Kaggle.
 - The "masks" dir will be automatically generated with the merged masks.
   Delete "masks" dir if you want to generate the merged masks again (ex. new data).
'''

DATA_DIR          = Path("data")
KAGGLE_TEST_DIR   = DATA_DIR / "kaggle" / "stage1_test"
KAGGLE_TRAIN_DIR  = DATA_DIR / "kaggle-dsbowl-2018-dataset-fixes" / "stage1_train"
KAGGLE_TRAIN_CSV  = DATA_DIR / "kaggle-dsbowl-2018-dataset-fixes" / "stage1_train_labels.csv"
MASKS_DIR         = DATA_DIR / "masks"


dataset    = nucleiDataset(KAGGLE_TRAIN_DIR, MASKS_DIR, CROP_SIZE, cfg.PIXEL_MEANS)
dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True, num_workers=num_workers, drop_last=drop_last_batch)
# TODO: Validation dataloader

train_size = len(dataset)
if drop_last_batch: number_of_batches = math.floor(train_size/batch_size)
else:               number_of_batches = math.ceil(train_size/batch_size)


###################################### DATA NEW


traindir = os.path.join(args.data, 'train')
valdir = os.path.join(args.data, 'val')
normalize = transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])

train_dataset = datasets.ImageFolder(
    traindir,
    transforms.Compose([
        transforms.RandomResizedCrop(224),
        transforms.RandomHorizontalFlip(),
        transforms.ToTensor(),
        normalize,
    ]))

if args.distributed:
    train_sampler = torch.utils.data.distributed.DistributedSampler(train_dataset)
else:
    train_sampler = None

train_loader = torch.utils.data.DataLoader(
    train_dataset, batch_size=args.batch_size, shuffle=(train_sampler is None),
    num_workers=args.workers, pin_memory=True, sampler=train_sampler)

val_loader = torch.utils.data.DataLoader(
    datasets.ImageFolder(valdir, transforms.Compose([
        transforms.Resize(256),
        transforms.CenterCrop(224),
        transforms.ToTensor(),
        normalize,
    ])),
    batch_size=args.batch_size, shuffle=False,
    num_workers=args.workers, pin_memory=True)



###################################### MODEL

# Select a torchvision model
if pretrained:
    print("=> using pre-trained model '{}'".format(architecture))
    model = torchvision.models.__dict__[architecture](pretrained=True)
else:
    print("=> creating model '{}'".format(architecture))
    model = torchvision.models.__dict__[architecture]()


if architecture.startswith('alexnet') or architecture.startswith('vgg'):
    model.features = torch.nn.DataParallel(model.features)
    model.cuda()
else:
    model = torch.nn.DataParallel(model).cuda()


#if multiple_gpus:
#    fasterRCNN = nn.DataParallel(fasterRCNN) # Prepare net to support multiple gpus
#
#if gpu:
#    fasterRCNN.cuda() # Load net to gpu (or gpus)



###################################### LOSS function (criterion) and optimizer

criterion = nn.CrossEntropyLoss().cuda()
optimizer = torch.optim.SGD(model.parameters(), learning_rate, momentum=momentum, weight_decay=weight_decay)

# optionally resume from a checkpoint
if resume:
    if os.path.isfile(resume):
        print("=> loading checkpoint '{}'".format(resume))
        checkpoint = torch.load(resume)
        args.start_epoch = checkpoint['epoch']
        best_prec1 = checkpoint['best_prec1']
        model.load_state_dict(checkpoint['state_dict'])
        optimizer.load_state_dict(checkpoint['optimizer'])
        print("=> loaded checkpoint '{}' (epoch {})"
              .format(resume, checkpoint['epoch']))
    else:
        print("=> no checkpoint found at '{}'".format(resume))

cudnn.benchmark = True




###################################### TRAIN NEW

def train(train_loader, model, criterion, optimizer, epoch):
    for epoch in range(args.start_epoch, args.epochs):
        if args.distributed:
            train_sampler.set_epoch(epoch)
        adjust_learning_rate(optimizer, epoch)

        # train for one epoch
        train_epoch(train_loader, model, criterion, optimizer, epoch)

        # evaluate on validation set
        prec1 = validate_epoch(val_loader, model, criterion)

        # remember best prec@1 and save checkpoint
        is_best = prec1 > best_prec1
        best_prec1 = max(prec1, best_prec1)
        save_checkpoint({
            'epoch': epoch + 1,
            'arch': args.arch,
            'state_dict': model.state_dict(),
            'best_prec1': best_prec1,
            'optimizer' : optimizer.state_dict(),
        }, is_best)


def train_epoch(train_loader, model, criterion, optimizer, epoch):
    batch_time = AverageMeter()
    data_time  = AverageMeter()
    losses     = AverageMeter()
    top1       = AverageMeter()
    top5       = AverageMeter()

    # switch to train mode
    model.train()

    end = time.time()
    for i, (input, target) in enumerate(train_loader):
        # measure data loading time
        data_time.update(time.time() - end)

        target = target.cuda(non_blocking=True)

        # FOWARD: compute output and loss
        output = model(input)
        loss = criterion(output, target)

        # measure accuracy and record loss
        prec1, prec5 = accuracy(output, target, topk=(1, 5))
        losses.update(loss.item(), input.size(0))
        top1.update(prec1[0], input.size(0))
        top5.update(prec5[0], input.size(0))

        # BACKWARD: compute gradient and do SGD step
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        # measure elapsed time
        batch_time.update(time.time() - end)
        end = time.time()

        if i % args.print_freq == 0:
            print('Epoch: [{0}][{1}/{2}]\t'
                  'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
                  'Data {data_time.val:.3f} ({data_time.avg:.3f})\t'
                  'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
                  'Prec@1 {top1.val:.3f} ({top1.avg:.3f})\t'
                  'Prec@5 {top5.val:.3f} ({top5.avg:.3f})'.format(
                   epoch, i, len(train_loader), batch_time=batch_time,
                   data_time=data_time, loss=losses, top1=top1, top5=top5))


def validate_epoch(val_loader, model, criterion):
    batch_time = AverageMeter()
    losses     = AverageMeter()
    top1       = AverageMeter()
    top5       = AverageMeter()

    # switch to evaluate mode
    model.eval()

    with torch.no_grad():
        end = time.time()
        for i, (input, target) in enumerate(val_loader):
            target = target.cuda(non_blocking=True)

            # FOWARD: compute output and loss
            output = model(input)
            loss = criterion(output, target)

            # measure accuracy and record loss
            prec1, prec5 = accuracy(output, target, topk=(1, 5))
            losses.update(loss.item(), input.size(0))
            top1.update(prec1[0], input.size(0))
            top5.update(prec5[0], input.size(0))

            # measure elapsed time
            batch_time.update(time.time() - end)
            end = time.time()

            if i % args.print_freq == 0:
                print('Test: [{0}/{1}]\t'
                      'Time {batch_time.val:.3f} ({batch_time.avg:.3f})\t'
                      'Loss {loss.val:.4f} ({loss.avg:.4f})\t'
                      'Prec@1 {top1.val:.3f} ({top1.avg:.3f})\t'
                      'Prec@5 {top5.val:.3f} ({top5.avg:.3f})'.format(
                       i, len(val_loader), batch_time=batch_time, loss=losses,
                       top1=top1, top5=top5))

        print(' * Prec@1 {top1.avg:.3f} Prec@5 {top5.avg:.3f}'
              .format(top1=top1, top5=top5))

    return top1.avg





def save_checkpoint(state, is_best, filename='checkpoint.pth.tar'):
    torch.save(state, filename)
    if is_best:
        shutil.copyfile(filename, 'model_best.pth.tar')


class AverageMeter(object):
    """Computes and stores the average and current value"""
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count


def adjust_learning_rate(optimizer, epoch):
    """Sets the learning rate to the initial LR decayed by 10 every 30 epochs"""
    lr = args.lr * (0.1 ** (epoch // 30))
    for param_group in optimizer.param_groups:
        param_group['lr'] = lr


def accuracy(output, target, topk=(1,)):
    """Computes the precision@k for the specified values of k"""
    with torch.no_grad():
        maxk = max(topk)
        batch_size = target.size(0)

        _, pred = output.topk(maxk, 1, True, True)
        pred = pred.t()
        correct = pred.eq(target.view(1, -1).expand_as(pred))

        res = []
        for k in topk:
            correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
            res.append(correct_k.mul_(100.0 / batch_size))
        return res
